Nested Joins acting incorrectly

ruanpepe opened the pull request:
If table A is linked to table B by a left join, and the table B is linked to table C by a inner join, when you try to access the table C from table A, the query acts like table A was linked to table B by a inner join.
I modified  the "_buildJoinPart" method to group nested joins by parentheses, and that is working fine now.

ruanpepe committed:
Nested Joins acting incorrectly

If table A is linked to table B by a left join, and the table B is linked to table C by a inner join, when you try to access the table C from table A, the query acts like table A was linked to table B by a inner join.
I modified  the "_buildJoinPart" method to group nested joins by parentheses, and that is working fine now.

stickler-ci committed:
Fixing style errors.

ruanpepe committed:
Doc commet add and using "count" instead "sizeof"

stickler-ci committed:
Fixing style errors.

markstory commented:
Any chance you could include a way to reproduce the problem you're having? Your fix has caused many tests to fail.

ruanpepe commented:
> Any chance you could include a way to reproduce the problem you're having? Your fix has caused many tests to fail.

My model:

JuridicalPeople model
`
class JuridicalPeopleTable extends Table
{
    public function initialize(array $config)
    {
        parent::initialize($config);

        $this->setTable('juridical_people');
        $this->setDisplayField('id');
        $this->setPrimaryKey('id');

        $this->hasOne('Teams', [
            'foreignKey' => 'juridical_person_id',
            'dependent' => true,
            'cascadeCallbacks' => true
        ]);
    }

    public function validationDefault(Validator $validator)
    {
        $validator
            ->integer('id')
            ->allowEmptyString('id', null, 'create');

        $validator
            ->scalar('logo')
            ->maxLength('logo', 255)
            ->allowEmptyString('logo');

        return $validator;
    }
}

class JuridicalPerson extends Entity
{
    protected $_accessible = [
        'logo' => true,
        'team' => true
    ];
}
`

My Teams model:
`
class TeamsTable extends Table
{
    public function initialize(array $config)
    {
        parent::initialize($config);

        $this->setTable('teams');
        $this->setDisplayField('id');
        $this->setPrimaryKey('id');

        $this->belongsTo('JuridicalPeople', [
            'foreignKey' => 'juridical_person_id',
            'joinType' => 'INNER'
        ]);
        
        $this->hasMany('Athletes', [
            'foreignKey' => 'team_id'
        ]);
    }

    public function validationDefault(Validator $validator)
    {
        $validator
            ->integer('id')
            ->allowEmptyString('id', null, 'create');

        return $validator;
    }

    public function buildRules(RulesChecker $rules)
    {
        $rules->add($rules->existsIn(['juridical_person_id'], 'JuridicalPeople'));

        return $rules;
    }
}

class Team extends Entity
{
    protected $_accessible = [
        'juridical_person_id' => true,
        'juridical_person' => true,
        'athletes' => true,
    ];
}
`

My Athletes model:
`
class AthletesTable extends Table
{
    public function initialize(array $config)
    {
        parent::initialize($config);

        $this->setTable('athletes');
        $this->setDisplayField('id');
        $this->setPrimaryKey('id');

        $this->belongsTo('Teams')
            ->setForeignKey('team_id');
    }

    public function validationDefault(Validator $validator)
    {
        $validator
            ->integer('id')
            ->allowEmptyString('id', null, 'create');

        return $validator;
    }

    public function buildRules(RulesChecker $rules)
    {
        $rules->add($rules->existsIn(['team_id'], 'Teams'));

        return $rules;
    }
}
`

My Athletes Controller:
`class AthletesController extends AppController
{
    public function index()
    {
        $this->paginate = [
            'contain' => [
                'NaturalPeople',
                'Teams.JuridicalPeople']
        ];
        
        $athletes = $this->paginate($this->Athletes->find());

        $this->set(compact('athletes'));
    }
}`

At my athletes/index, if the athlete doesn't match with a team, he should be showed becouse the Athletes table is joined to Teams table by a left join. But this is not happening becouse the teams table is joined to JuridicalPeople table by inner join, and the current way to build query insert all joins at same level. So, at query, the Athletes table is joined to the JuridicalPeople table by a inner join.

ruanpepe committed:
Prevent repeated joins

The code change was causing a error of Repeated Alias at same querys. Fixed at line 247.

stickler-ci committed:
Fixing style errors.

othercorey commented:
@ruanpepe Your editor is changing every coding standard.  Please adjust it to stop.

markstory commented on src/Database/QueryCompiler.php:
```diff
@@ -234,12 +256,17 @@ protected function _buildJoinPart($parts, $query, $generator)
             }
 
-            $joins .= sprintf(' %s JOIN %s %s', $join['type'], $join['table'], $join['alias']);
+            $joins .= sprintf(' %s JOIN (%s %s', $join['type'], $join['table'], $join['alias']);
```
Could we only use parenthesis on join expressions that contain child associations? I think this change is why so many tests are failing.

ruanpepe replied:
> Could we only use parenthesis on join expressions that contain child associations? I think this change is why so many tests are failing.

Yes. Te query join structure was changed from
```
LEFT JOIN teams Teams 
ON Teams.id = (Athletes.team_id) 

INNER JOIN juridical_people JuridicalPeople 
ON JuridicalPeople.id = (Teams.juridical_person_id)
```
to
```
LEFT JOIN (teams Teams 
	INNER JOIN (juridical_people JuridicalPeople) 
	ON JuridicalPeople.id = (Teams.juridical_person_id)
) 
ON Teams.id = (Athletes.team_id) 
```
and it's working fine to me.
Another way to do this, would be inserting a subquery, 
```
LEFT JOIN 
(
	SELECT 
		...
		...
	FROM teams Teams
	INNER JOIN juridical_people JuridicalPeople
	ON JuridicalPeople.id = Teams.juridical_person_id
) Teams  
ON Teams.id = (Athletes.team_id) 
```
but i think it would cause many performance troubles.

ruanpepe committed:
Some joins were not being loaded

stickler-ci committed:
Fixing style errors.

ruanpepe committed:
Update QueryCompiler.php

stickler-ci committed:
Fixing style errors.

othercorey commented:
We are going to approach this a different way in 4.3.
